// @novu/react@3.16.0/dist/cjs/index.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var index_exports = {};
__export(index_exports, {
  Bell: () => import_components.Bell,
  Inbox: () => import_components.Inbox,
  InboxContent: () => import_components.InboxContent,
  Notifications: () => import_components.Notifications,
  NovuProvider: () => import_components.NovuProvider,
  PreferenceLevel: () => import_js.PreferenceLevel,
  Preferences: () => import_components.Preferences,
  SeverityLevelEnum: () => import_js.SeverityLevelEnum,
  SlackConnectButton: () => import_components.SlackConnectButton,
  SlackLinkUser: () => import_components.SlackLinkUser,
  Subscription: () => import_components.Subscription,
  SubscriptionButton: () => import_components.SubscriptionButton,
  SubscriptionPreferences: () => import_components.SubscriptionPreferences,
  WorkflowCriticalityEnum: () => import_js.WorkflowCriticalityEnum,
  useChannelConnection: () => import_hooks.useChannelConnection,
  useChannelConnections: () => import_hooks.useChannelConnections,
  useChannelEndpoint: () => import_hooks.useChannelEndpoint,
  useCounts: () => import_hooks.useCounts,
  useCreateChannelEndpoint: () => import_hooks.useCreateChannelEndpoint,
  useCreateSubscription: () => import_hooks.useCreateSubscription,
  useDeleteChannelEndpoint: () => import_hooks.useDeleteChannelEndpoint,
  useNotifications: () => import_hooks.useNotifications,
  useNovu: () => import_hooks.useNovu,
  usePreferences: () => import_hooks.usePreferences,
  useRemoveSubscription: () => import_hooks.useRemoveSubscription,
  useSchedule: () => import_hooks.useSchedule,
  useSubscription: () => import_hooks.useSubscription,
  useSubscriptions: () => import_hooks.useSubscriptions,
  useUpdateSubscription: () => import_hooks.useUpdateSubscription
});
module.exports = __toCommonJS(index_exports);
var import_js = require("@novu/js");
var import_components = require("./components/index.cjs");
var import_hooks = require("./hooks/index.cjs");
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Bell,
  Inbox,
  InboxContent,
  Notifications,
  NovuProvider,
  PreferenceLevel,
  Preferences,
  SeverityLevelEnum,
  SlackConnectButton,
  SlackLinkUser,
  Subscription,
  SubscriptionButton,
  SubscriptionPreferences,
  WorkflowCriticalityEnum,
  useChannelConnection,
  useChannelConnections,
  useChannelEndpoint,
  useCounts,
  useCreateChannelEndpoint,
  useCreateSubscription,
  useDeleteChannelEndpoint,
  useNotifications,
  useNovu,
  usePreferences,
  useRemoveSubscription,
  useSchedule,
  useSubscription,
  useSubscriptions,
  useUpdateSubscription
});
