// @novu/react@3.16.0/dist/cjs/hooks/NovuProvider.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/hooks/NovuProvider.tsx
var NovuProvider_exports = {};
__export(NovuProvider_exports, {
  InternalNovuProvider: () => InternalNovuProvider,
  NovuProvider: () => NovuProvider,
  useNovu: () => useNovu,
  useRealtime: () => useRealtime,
  useUnsafeNovu: () => useUnsafeNovu
});
module.exports = __toCommonJS(NovuProvider_exports);
var import_js = require("@novu/js");
var import_internal = require("@novu/js/internal");
var import_react = require("react");
var import_jsx_runtime = require("react/jsx-runtime");
var NovuContext = (0, import_react.createContext)({ novu: void 0, realtime: true });
var NovuProvider = (props) => {
  const { subscriberId, realtime = true, ...propsWithoutSubscriberId } = props;
  const subscriberObj = (0, import_react.useMemo)(
    () => (0, import_internal.buildSubscriber)({ subscriberId, subscriber: props.subscriber }),
    [subscriberId, props.subscriber]
  );
  const applicationIdentifier = propsWithoutSubscriberId.applicationIdentifier ? propsWithoutSubscriberId.applicationIdentifier : "";
  const providerProps = {
    ...propsWithoutSubscriberId,
    applicationIdentifier,
    subscriber: subscriberObj,
    realtime
  };
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(InternalNovuProvider, { ...providerProps, applicationIdentifier, children: props.children });
};
var InternalNovuProvider = (props) => {
  const applicationIdentifier = props.applicationIdentifier || "";
  const subscriberObj = (0, import_react.useMemo)(
    () => (0, import_internal.buildSubscriber)({ subscriberId: props.subscriberId, subscriber: props.subscriber }),
    [props.subscriberId, props.subscriber]
  );
  const {
    children,
    subscriberHash,
    contextHash,
    backendUrl,
    apiUrl,
    socketUrl,
    socketOptions,
    useCache,
    defaultSchedule,
    context,
    realtime = true
  } = props;
  const value = (0, import_react.useMemo)(
    () => ({
      novu: new import_js.Novu({
        applicationIdentifier,
        subscriberHash,
        contextHash,
        backendUrl,
        apiUrl,
        socketUrl,
        socketOptions,
        useCache,
        subscriber: subscriberObj,
        defaultSchedule,
        context
      }),
      realtime
    }),
    [
      applicationIdentifier,
      subscriberHash,
      subscriberObj,
      context,
      contextHash,
      backendUrl,
      apiUrl,
      socketUrl,
      socketOptions,
      useCache,
      realtime
    ]
  );
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(NovuContext.Provider, { value, children });
};
var useNovu = () => {
  const context = (0, import_react.useContext)(NovuContext);
  if (!context.novu) {
    throw new Error("useNovu must be used within a <NovuProvider />");
  }
  return context.novu;
};
var useUnsafeNovu = () => {
  const context = (0, import_react.useContext)(NovuContext);
  return context.novu;
};
var useRealtime = () => {
  const context = (0, import_react.useContext)(NovuContext);
  return context.realtime;
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  InternalNovuProvider,
  NovuProvider,
  useNovu,
  useRealtime,
  useUnsafeNovu
});
