// @djpanda/convex-authz@2.3.1/dist/client/index.js
/**
 * @djpanda/convex-authz - Authorization Component for Convex
 *
 * A comprehensive RBAC/ABAC/ReBAC authorization component featuring
 * O(1) indexed lookups, inspired by Google Zanzibar.
 *
 * @example
 * ```typescript
 * import { Authz, definePermissions, defineRoles } from "@djpanda/convex-authz";
 * import { components } from "./_generated/api";
 *
 * const permissions = definePermissions({
 *   documents: { create: true, read: true, update: true, delete: true },
 * });
 *
 * const roles = defineRoles(permissions, {
 *   admin: { documents: ["create", "read", "update", "delete"] },
 *   viewer: { documents: ["read"] },
 * });
 *
 * export const authz = new Authz(components.authz, { permissions, roles, tenantId: "my-app" });
 * ```
 */
import { ConvexError } from "convex/values";
import { validateTenantId, validateUserId, validatePermission, validateScope, validateRole, validateOptionalExpiresAt, validateAttributeKey, validateAuditLimit, validateRelationArgs, validatePermissions, validateRoleAssignItems, validateRoles, } from "./validation.js";
/** Reserved keys in role definitions; do not use as permission resource names. */
const RESERVED_ROLE_KEYS = ["inherits", "includes"];
export function definePermissions(...definitions) {
    if (definitions.length === 1)
        return definitions[0];
    const result = {};
    for (const def of definitions) {
        for (const [resource, actions] of Object.entries(def)) {
            result[resource] = { ...(result[resource] ?? {}), ...actions };
        }
    }
    return result;
}
export function defineRoles(_permissions, ...roleDefs) {
    if (roleDefs.length === 1)
        return roleDefs[0];
    const result = {};
    for (const def of roleDefs) {
        for (const [roleName, rolePerms] of Object.entries(def)) {
            const r = rolePerms;
            if (!result[roleName]) {
                result[roleName] = { ...r };
            }
            else {
                const existing = result[roleName];
                if (r.inherits !== undefined)
                    existing.inherits = r.inherits;
                if (Array.isArray(r.includes)) {
                    const prev = existing.includes ?? [];
                    existing.includes = [...new Set([...prev, ...r.includes])];
                }
                for (const [key, value] of Object.entries(r)) {
                    if (RESERVED_ROLE_KEYS.includes(key))
                        continue;
                    const existingArr = (existing[key] ?? []);
                    const incomingArr = Array.isArray(value) ? value : [];
                    existing[key] = [...new Set([...existingArr, ...incomingArr])];
                }
            }
        }
    }
    return result;
}
/**
 * Define ABAC policies
 */
export function definePolicies(policies) {
    return policies;
}
/**
 * Evaluate a policy condition (sync or async). Always returns a Promise so callers can await uniformly.
 */
export function evaluatePolicyCondition(condition, ctx) {
    try {
        return Promise.resolve(condition(ctx)).catch(() => false);
    }
    catch {
        return Promise.resolve(false);
    }
}
/**
 * Resolve effective permissions for a role, following inherits and includes with cycle detection.
 * @internal
 */
function resolveRolePermissions(roles, roleName, visited) {
    if (!Object.prototype.hasOwnProperty.call(roles, roleName))
        return [];
    if (visited.has(roleName)) {
        throw new Error(`Role inheritance cycle detected involving role "${roleName}"`);
    }
    visited.add(roleName);
    const rolePerms = roles[roleName];
    const perms = new Set();
    try {
        const inherits = rolePerms.inherits;
        if (inherits !== undefined && inherits !== null) {
            const ref = String(inherits);
            if (!Object.prototype.hasOwnProperty.call(roles, ref)) {
                throw new Error(`Role "${roleName}" inherits unknown role "${ref}"`);
            }
            for (const p of resolveRolePermissions(roles, ref, visited))
                perms.add(p);
        }
        const includes = rolePerms.includes;
        if (Array.isArray(includes)) {
            for (const ref of includes) {
                const r = String(ref);
                if (!Object.prototype.hasOwnProperty.call(roles, r)) {
                    throw new Error(`Role "${roleName}" includes unknown role "${r}"`);
                }
                for (const p of resolveRolePermissions(roles, r, visited))
                    perms.add(p);
            }
        }
        for (const [resource, actions] of Object.entries(rolePerms)) {
            if (RESERVED_ROLE_KEYS.includes(resource))
                continue;
            if (Array.isArray(actions)) {
                for (const action of actions)
                    perms.add(`${resource}:${String(action)}`);
            }
        }
        return [...perms];
    }
    finally {
        visited.delete(roleName);
    }
}
/**
 * Flatten role permissions into an array of permission strings.
 * Resolves inheritance (`inherits`) and composition (`includes`) so that
 * effective permissions include those from inherited/included roles.
 */
export function flattenRolePermissions(roles, roleName) {
    const visited = new Set();
    return resolveRolePermissions(roles, roleName, visited);
}
// ============================================================================
// Authz Client Class (Standard)
// ============================================================================
/**
 * Standard Authz client for RBAC/ABAC operations
 *
 * @example
 * ```typescript
 * const authz = new Authz(components.authz, { permissions, roles, tenantId: "my-app" });
 *
 * // In a mutation or query
 * const canEdit = await authz.can(ctx, userId, "documents:update");
 * await authz.require(ctx, userId, "documents:update");
 * ```
 */
export class Authz {
    component;
    options;
    constructor(component, options) {
        this.component = component;
        this.options = options;
        validateTenantId(options.tenantId);
    }
    withTenant(tenantId) {
        validateTenantId(tenantId);
        return new Authz(this.component, { ...this.options, tenantId });
    }
    /**
     * Build role permissions map for queries
     */
    buildRolePermissionsMap() {
        const map = {};
        const roles = this.options.roles;
        for (const roleName of Object.keys(roles)) {
            map[roleName] = flattenRolePermissions(roles, roleName);
        }
        return map;
    }
    /**
     * Internal helper: check permission via the unified O(1) indexed path.
     * Returns the full structured result including tier and policyName for deferred evaluation.
     */
    async _checkPermission(ctx, userId, permission, scope) {
        validateUserId(userId);
        validatePermission(permission);
        validateScope(scope);
        return ctx.runQuery(this.component.unified.checkPermission, {
            tenantId: this.options.tenantId,
            userId,
            permission,
            scope,
        });
    }
    /**
     * Internal helper: evaluate a deferred policy condition.
     * Fetches user attributes and roles, builds a PolicyContext, and evaluates the condition.
     */
    async _evaluateDeferredPolicy(ctx, userId, policyName, permission, scope, requestContext) {
        if (!policyName || !this.options.policies)
            return false;
        const policy = this.options.policies[policyName];
        if (!policy)
            return false;
        // Fetch user attributes and roles for the context
        const [attrs, roles] = await Promise.all([
            ctx.runQuery(this.component.queries.getUserAttributes, {
                userId,
                tenantId: this.options.tenantId,
            }),
            ctx.runQuery(this.component.queries.getUserRoles, {
                userId,
                scope,
                tenantId: this.options.tenantId,
            }),
        ]);
        const roleNames = roles.map((r) => r.role);
        const attrMap = Object.fromEntries(attrs.map((a) => [a.key, a.value]));
        const policyCtx = {
            subject: {
                userId,
                roles: roleNames,
                attributes: attrMap,
            },
            resource: scope ? { type: scope.type, id: scope.id, ...requestContext } : requestContext ? { type: "", id: "", ...requestContext } : undefined,
            action: permission,
            environment: { timestamp: Date.now() },
            hasRole: (role) => roleNames.includes(role),
            hasAttribute: (key) => key in attrMap,
            getAttribute: (key, defaultValue) => attrMap[key] ?? defaultValue,
        };
        return evaluatePolicyCondition(policy.condition, policyCtx);
    }
    /**
     * Check if user has permission.
     * Uses O(1) indexed lookup via the unified checkPermission query.
     * Permission checks support wildcard matching: if the user has a role or override with a pattern
     * (e.g. "documents:*", "*:read", or "*"), it is treated as granting that permission when the
     * pattern matches the requested permission.
     * @param permission - Concrete permission "resource:action" (e.g. "documents:read")
     */
    async can(ctx, userId, permission, scope) {
        const result = await this._checkPermission(ctx, userId, permission, scope);
        if (!result.allowed)
            return false;
        // For deferred policies, evaluate with empty context
        if (result.tier === "deferred" && this.options.policies) {
            return this._evaluateDeferredPolicy(ctx, userId, result.policyName, permission, scope);
        }
        return true;
    }
    /**
     * Check if user has permission, with additional request context for deferred policy evaluation.
     * Uses O(1) indexed lookup via the unified checkPermission query.
     * @param permission - Concrete permission "resource:action" (e.g. "documents:read")
     * @param requestContext - Additional context passed to deferred policy conditions
     */
    async canWithContext(ctx, userId, permission, scope, requestContext) {
        const result = await this._checkPermission(ctx, userId, permission, scope);
        if (!result.allowed)
            return false;
        if (result.tier === "deferred" && this.options.policies) {
            return this._evaluateDeferredPolicy(ctx, userId, result.policyName, permission, scope, requestContext);
        }
        return true;
    }
    /**
     * Require permission or throw error.
     * Supports the same wildcard matching as {@link Authz.can}.
     * @param permission - Concrete permission "resource:action" (e.g. "documents:read")
     */
    async require(ctx, userId, permission, scope) {
        const allowed = await this.can(ctx, userId, permission, scope);
        if (!allowed) {
            // ConvexError (not plain Error) so the rejection is classified as an
            // expected, structured failure on the client side instead of an
            // "Uncaught Error / Server Error" with internal stack trace exposed.
            // Consumers can catch and pattern-match on `.data.code === "FORBIDDEN"`.
            throw new ConvexError({
                code: "FORBIDDEN",
                message: `Permission denied: ${permission}${scope ? ` on ${scope.type}:${scope.id}` : ""}`,
                permission: String(permission),
                scopeType: scope?.type ?? null,
                scopeId: scope?.id ?? null,
            });
        }
    }
    /**
     * Check if user has any of the given permissions (canAny).
     * Returns true if the user is allowed at least one of the permissions in the given scope.
     * @param permissions - Array of permission strings (e.g. ["documents:read", "documents:update"]). Max length 100.
     */
    async canAny(ctx, userId, permissions, scope) {
        validateUserId(userId);
        validatePermissions(permissions);
        validateScope(scope);
        for (const perm of permissions) {
            if (await this.can(ctx, userId, perm, scope)) {
                return true;
            }
        }
        return false;
    }
    /**
     * Check if user has a role - O(1) indexed lookup
     */
    async hasRole(ctx, userId, role, scope) {
        validateUserId(userId);
        validateRole(role, this.options.roles);
        validateScope(scope);
        return await ctx.runQuery(this.component.indexed.hasRoleFast, {
            userId,
            role,
            objectType: scope?.type,
            objectId: scope?.id,
            tenantId: this.options.tenantId,
        });
    }
    /**
     * Get all roles for a user - O(1) indexed lookup
     */
    async getUserRoles(ctx, userId, scope) {
        validateUserId(userId);
        validateScope(scope);
        const scopeKey = scope ? `${scope.type}:${scope.id}` : undefined;
        return await ctx.runQuery(this.component.indexed.getUserRolesFast, {
            userId,
            scopeKey,
            tenantId: this.options.tenantId,
        });
    }
    /**
     * Get all effective permissions for a user - O(1) indexed lookup
     */
    async getUserPermissions(ctx, userId, scope) {
        validateUserId(userId);
        validateScope(scope);
        const scopeKey = scope ? `${scope.type}:${scope.id}` : undefined;
        return await ctx.runQuery(this.component.indexed.getUserPermissionsFast, {
            userId,
            scopeKey,
            tenantId: this.options.tenantId,
        });
    }
    /**
     * Get all attributes for a user
     */
    async getUserAttributes(ctx, userId) {
        validateUserId(userId);
        return await ctx.runQuery(this.component.queries.getUserAttributes, {
            userId,
            tenantId: this.options.tenantId,
        });
    }
    /**
     * Assign a role to a user (unified: writes to both source tables and indexed tables)
     */
    async assignRole(ctx, userId, role, scope, expiresAt, actorId) {
        validateUserId(userId);
        validateRole(role, this.options.roles);
        validateScope(scope);
        validateOptionalExpiresAt(expiresAt);
        const rolePermissions = flattenRolePermissions(this.options.roles, role);
        // Build policy classifications for permissions being assigned
        // Static policies (type: "static" or undefined) are classified as "deferred" because
        // we cannot evaluate them at assignment time without user context (MutationCtx lacks runQuery).
        // They will be properly evaluated at read time in can().
        const policyClassifications = {};
        if (this.options.policies) {
            for (const perm of rolePermissions) {
                const policy = this.options.policies[perm];
                if (policy) {
                    policyClassifications[perm] = "deferred";
                }
            }
        }
        return await ctx.runMutation(this.component.unified.assignRoleUnified, {
            tenantId: this.options.tenantId,
            userId,
            role,
            rolePermissions,
            scope,
            expiresAt,
            assignedBy: actorId ?? this.options.defaultActorId,
            enableAudit: true,
            policyClassifications: Object.keys(policyClassifications).length > 0 ? policyClassifications : undefined,
        });
    }
    /**
     * Revoke a role from a user (unified: updates both source tables and indexed tables)
     */
    async revokeRole(ctx, userId, role, scope, actorId) {
        validateUserId(userId);
        validateRole(role, this.options.roles);
        validateScope(scope);
        const rolePermissions = flattenRolePermissions(this.options.roles, role);
        return await ctx.runMutation(this.component.unified.revokeRoleUnified, {
            tenantId: this.options.tenantId,
            userId,
            role,
            rolePermissions,
            scope,
            revokedBy: actorId ?? this.options.defaultActorId,
            enableAudit: true,
        });
    }
    /**
     * Assign multiple roles to a user in a single transaction.
     * @param roles - Array of { role, scope?, expiresAt?, metadata? }. Max length 100.
     */
    async assignRoles(ctx, userId, roles, actorId) {
        validateUserId(userId);
        validateRoleAssignItems(roles, this.options.roles);
        const assignedBy = actorId ?? this.options.defaultActorId;
        // Build policy classifications from all permissions across all roles being assigned
        // All policies (static and deferred) are classified as "deferred" because
        // we cannot evaluate static policies at assignment time without user context.
        const policyClassifications = {};
        if (this.options.policies) {
            const rolePermissionsMap = this.buildRolePermissionsMap();
            for (const r of roles) {
                const perms = rolePermissionsMap[r.role] ?? [];
                for (const perm of perms) {
                    const policy = this.options.policies[perm];
                    if (policy) {
                        policyClassifications[perm] = "deferred";
                    }
                }
            }
        }
        return ctx.runMutation(this.component.unified.assignRolesUnified, {
            userId,
            roles: roles.map((r) => ({
                role: r.role,
                scope: r.scope,
                expiresAt: r.expiresAt,
                metadata: r.metadata,
            })),
            rolePermissionsMap: this.buildRolePermissionsMap(),
            assignedBy,
            enableAudit: true,
            tenantId: this.options.tenantId,
            policyClassifications: Object.keys(policyClassifications).length > 0 ? policyClassifications : undefined,
        });
    }
    /**
     * Revoke multiple roles from a user in a single transaction.
     * @param roles - Array of { role, scope? }. Max length 100.
     */
    async revokeRoles(ctx, userId, roles, actorId) {
        validateUserId(userId);
        validateRoles(roles, this.options.roles);
        return ctx.runMutation(this.component.unified.revokeRolesUnified, {
            userId,
            roles: roles.map((r) => ({ role: r.role, scope: r.scope })),
            rolePermissionsMap: this.buildRolePermissionsMap(),
            revokedBy: actorId ?? this.options.defaultActorId,
            enableAudit: true,
            tenantId: this.options.tenantId,
        });
    }
    /**
     * Revoke all roles from a user (optionally only in a given scope).
     * Use for bulk cleanup or partial offboarding.
     */
    async revokeAllRoles(ctx, userId, scope, actorId) {
        validateUserId(userId);
        validateScope(scope);
        return ctx.runMutation(this.component.unified.revokeAllRolesUnified, {
            userId,
            scope,
            rolePermissionsMap: this.buildRolePermissionsMap(),
            revokedBy: actorId ?? this.options.defaultActorId,
            enableAudit: true,
            tenantId: this.options.tenantId,
        });
    }
    /**
     * Full user offboarding: remove all roles, permission overrides, attributes, and optionally
     * ReBAC relationships for the user (optionally scoped). Also clears indexed
     * effectiveRoles/effectivePermissions/effectiveRelationships when present.
     */
    async offboardUser(ctx, userId, options) {
        validateUserId(userId);
        if (options?.scope)
            validateScope(options.scope);
        return await ctx.runMutation(this.component.mutations.offboardUser, {
            userId,
            scope: options?.scope,
            revokedBy: options?.actorId ?? this.options.defaultActorId,
            removeAttributes: options?.removeAttributes,
            removeOverrides: options?.removeOverrides,
            removeRelationships: options?.removeRelationships,
            enableAudit: true,
            tenantId: this.options.tenantId,
        });
    }
    /**
     * Full user deprovisioning: wipes all roles, attributes, relationships, and permission
     * overrides for a given userId in one atomic call. Use for security incident response,
     * enterprise offboarding, or single-button deactivation.
     */
    async deprovisionUser(ctx, userId, options) {
        validateUserId(userId);
        return await ctx.runMutation(this.component.mutations.deprovisionUser, {
            userId,
            revokedBy: options?.actorId ?? this.options.defaultActorId,
            enableAudit: options?.enableAudit ?? true,
            tenantId: this.options.tenantId,
        });
    }
    /**
     * Set a user attribute
     */
    async setAttribute(ctx, userId, key, value, actorId) {
        validateUserId(userId);
        validateAttributeKey(key);
        return await ctx.runMutation(this.component.unified.setAttributeWithRecompute, {
            tenantId: this.options.tenantId,
            userId,
            key,
            value,
            setBy: actorId ?? this.options.defaultActorId,
            enableAudit: true,
        });
    }
    /**
     * Remove a user attribute
     */
    async removeAttribute(ctx, userId, key, actorId) {
        validateUserId(userId);
        validateAttributeKey(key);
        const result = await ctx.runMutation(this.component.mutations.removeAttribute, {
            userId,
            key,
            removedBy: actorId ?? this.options.defaultActorId,
            enableAudit: true,
            tenantId: this.options.tenantId,
        });
        // Trigger policy re-evaluation after attribute removal (policies may depend on this attribute)
        // All policies are classified as "deferred" so they are re-evaluated at read time.
        if (this.options.policies && Object.keys(this.options.policies).length > 0) {
            const policyClassifications = {};
            for (const [name] of Object.entries(this.options.policies)) {
                policyClassifications[name] = "deferred";
            }
            await ctx.runMutation(this.component.unified.recomputeUser, {
                tenantId: this.options.tenantId,
                userId,
                rolePermissionsMap: this.buildRolePermissionsMap(),
                policyClassifications,
            });
        }
        return result;
    }
    /**
     * Grant a direct permission override (unified: writes to both source tables and indexed tables).
     * Permission can be a concrete permission ("documents:read") or a wildcard pattern
     * ("documents:*", "*:read", "*:*", or "*") to allow all matching permissions.
     * @param permission - Permission or pattern (e.g. "documents:read", "documents:*", "*:read", "*")
     */
    async grantPermission(ctx, userId, permission, scope, reason, expiresAt, actorId) {
        validateUserId(userId);
        validatePermission(permission);
        validateScope(scope);
        validateOptionalExpiresAt(expiresAt);
        return await ctx.runMutation(this.component.unified.grantPermissionUnified, {
            tenantId: this.options.tenantId,
            userId,
            permission,
            scope,
            reason,
            expiresAt,
            createdBy: actorId ?? this.options.defaultActorId,
            enableAudit: true,
        });
    }
    /**
     * Deny a permission (explicit deny override, unified: writes to both source tables and indexed tables).
     * Permission can be a concrete permission or a wildcard pattern (e.g. "documents:*", "*:read", "*").
     * @param permission - Permission or pattern to deny
     */
    async denyPermission(ctx, userId, permission, scope, reason, expiresAt, actorId) {
        validateUserId(userId);
        validatePermission(permission);
        validateScope(scope);
        validateOptionalExpiresAt(expiresAt);
        return await ctx.runMutation(this.component.unified.denyPermissionUnified, {
            tenantId: this.options.tenantId,
            userId,
            permission,
            scope,
            reason,
            expiresAt,
            createdBy: actorId ?? this.options.defaultActorId,
            enableAudit: true,
        });
    }
    /**
     * Check relationship - O(1) indexed lookup
     */
    async hasRelation(ctx, subject, relation, object) {
        return ctx.runQuery(this.component.indexed.hasRelationFast, {
            subjectType: subject.type,
            subjectId: subject.id,
            relation,
            objectType: object.type,
            objectId: object.id,
            tenantId: this.options.tenantId,
        });
    }
    /**
     * Add a relationship (unified: writes to both source tables and indexed tables)
     */
    async addRelation(ctx, subject, relation, object, options) {
        validateRelationArgs(subject.type, subject.id, relation, object.type, object.id);
        return ctx.runMutation(this.component.unified.addRelationUnified, {
            subjectType: subject.type,
            subjectId: subject.id,
            relation,
            objectType: object.type,
            objectId: object.id,
            caveat: options?.caveat,
            caveatContext: options?.caveatContext,
            createdBy: options?.createdBy ?? this.options.defaultActorId,
            enableAudit: true,
            tenantId: this.options.tenantId,
            relationPermissions: this.options.relationPermissions,
        });
    }
    /**
     * Remove a relationship (unified: removes from both source tables and indexed tables)
     */
    async removeRelation(ctx, subject, relation, object, actorId) {
        validateRelationArgs(subject.type, subject.id, relation, object.type, object.id);
        return ctx.runMutation(this.component.unified.removeRelationUnified, {
            subjectType: subject.type,
            subjectId: subject.id,
            relation,
            objectType: object.type,
            objectId: object.id,
            removedBy: actorId ?? this.options.defaultActorId,
            enableAudit: true,
            tenantId: this.options.tenantId,
            relationPermissions: this.options.relationPermissions,
        });
    }
    /**
     * Build policy classifications from the configured policies. All policies
     * are classified as "deferred" so they are re-evaluated at read time.
     * Returns `undefined` when no policies are configured (matches the
     * optional argument shape on the component-side mutations).
     */
    buildPolicyClassifications() {
        if (!this.options.policies)
            return undefined;
        const entries = Object.entries(this.options.policies);
        if (entries.length === 0)
            return undefined;
        const map = {};
        for (const [name] of entries) {
            map[name] = "deferred";
        }
        return map;
    }
    /**
     * Recompute all indexed data for a user (effectiveRoles, effectivePermissions).
     * Useful when role definitions change and you need to rebuild the index.
     */
    async recomputeUser(ctx, userId) {
        validateUserId(userId);
        await ctx.runMutation(this.component.unified.recomputeUser, {
            tenantId: this.options.tenantId,
            userId,
            rolePermissionsMap: this.buildRolePermissionsMap(),
            policyClassifications: this.buildPolicyClassifications(),
        });
    }
    /**
     * Re-materialize permissions for every user holding the given role.
     *
     * Use after a role definition (`defineRoles(...)`) gains or loses
     * permissions and you redeploy. Existing assignments still carry the old
     * materialized rows in `effectivePermissions`; this rebuilds them from the
     * current role definition while preserving direct grants/denies.
     *
     * Requires an action context — internally pages through assignments and
     * runs `recomputeUser` per user. Each user is one mutation transaction;
     * a partial failure leaves earlier users updated and stops on the failing
     * one (rerun is safe and idempotent).
     *
     * @param role - Role name from the configured catalog. Type-checked.
     * @returns Number of unique users recomputed.
     */
    async syncRole(ctx, role) {
        const roles = this.options.roles;
        if (!(role in roles)) {
            throw new Error(`Role "${String(role)}" not found in role catalog`);
        }
        return await ctx.runAction(this.component.unified.syncRoleAction, {
            tenantId: this.options.tenantId,
            role,
            rolePermissionsMap: this.buildRolePermissionsMap(),
            policyClassifications: this.buildPolicyClassifications(),
        });
    }
    /**
     * Re-materialize permissions for every user holding any role in the
     * configured catalog. Convenience wrapper around `syncRole` that iterates
     * across all roles.
     *
     * Cost is roughly `(unique users with any role) × (one mutation each)`.
     * For a full deploy-time sync of a 10k-user app, expect tens of seconds.
     * For very large user bases, prefer running per-role syncs in parallel
     * (each `syncRole` call is independent).
     */
    async syncRoles(ctx) {
        const rolePermissionsMap = this.buildRolePermissionsMap();
        const policyClassifications = this.buildPolicyClassifications();
        const roles = Object.keys(rolePermissionsMap);
        let totalUsers = 0;
        for (const role of roles) {
            const result = await ctx.runAction(this.component.unified.syncRoleAction, {
                tenantId: this.options.tenantId,
                role,
                rolePermissionsMap,
                policyClassifications,
            });
            totalUsers += result.usersProcessed;
        }
        return { rolesProcessed: roles.length, usersProcessed: totalUsers };
    }
    /**
     * Get audit log entries.
     * Pass cursor and numItems for cursor-based pagination (returns { page, isDone, continueCursor }).
     * Omit both for legacy behavior (returns array, optional limit).
     */
    async getAuditLog(ctx, options) {
        if (options?.userId !== undefined)
            validateUserId(options.userId);
        if (options?.limit !== undefined)
            validateAuditLimit(options.limit);
        if (options?.numItems !== undefined)
            validateAuditLimit(options.numItems);
        const usePagination = options?.numItems !== undefined || options?.cursor !== undefined;
        const paginationOpts = usePagination
            ? {
                numItems: options?.numItems ?? 100,
                cursor: options?.cursor ?? null,
            }
            : undefined;
        return await ctx.runQuery(this.component.queries.getAuditLog, {
            userId: options?.userId,
            action: options?.action,
            limit: options?.limit,
            paginationOpts,
            tenantId: this.options.tenantId,
        });
    }
}
// ============================================================================
// v2 Definition Helpers
// ============================================================================
/**
 * Define traversal rules for ReBAC relation-graph walks.
 * This is a type-safe identity helper — it returns the value unchanged.
 *
 * @example
 * ```ts
 * const traversalRules = defineTraversalRules({
 *   user: [{ through: "member", via: "group", inherit: "viewer" }],
 * });
 * ```
 */
export function defineTraversalRules(rules) {
    return rules;
}
/**
 * Define a map from relation names to the permission strings they grant.
 * This is a type-safe identity helper — it returns the value unchanged.
 *
 * @example
 * ```ts
 * const relationPermissions = defineRelationPermissions({
 *   owner: ["documents:read", "documents:write"],
 *   viewer: ["documents:read"],
 * });
 * ```
 */
export function defineRelationPermissions(map) {
    return map;
}
/**
 * Define named caveat functions for fine-grained permission conditions.
 * This is a type-safe identity helper — it returns the value unchanged.
 *
 * @example
 * ```ts
 * const caveats = defineCaveats({
 *   isOwner: ({ subject, object }) => subject.id === object.id,
 * });
 * ```
 */
export function defineCaveats(caveats) {
    return caveats;
}
/**
 * Check if a concrete permission matches a permission pattern (supports wildcards).
 * Use this for client-side logic when you need to test whether a stored pattern
 * would grant a given permission.
 *
 * Patterns:
 * - `"*"` matches every permission
 * - `"resource:*"` matches all actions on that resource (e.g. `"documents:*"` matches `documents:read`, `documents:write`)
 * - `"*:action"` matches that action on any resource (e.g. `"*:read"` matches `documents:read`, `settings:read`)
 * - `"*:*"` matches every permission
 *
 * @example
 * ```ts
 * import { matchesPermissionPattern } from "@djpanda/convex-authz";
 * matchesPermissionPattern("documents:read", "documents:*"); // true
 * matchesPermissionPattern("documents:read", "*:read");     // true
 * matchesPermissionPattern("documents:write", "*:read");   // false
 * ```
 */
export { matchesPermissionPattern, parsePermission, buildPermission, } from "../component/helpers.js";
